Colored list #4

Matt3o12 opened the pull request:
## Colored lists #4 ##

**Don't merge, yet!**
*The development has not quite finished, yet. This is just a preview.*

![screen shot 2015-04-06 at 12 15 21 am](https://cloud.githubusercontent.com/assets/1218474/6998927/1756cf40-dbf2-11e4-994a-30c6f4e50e97.png)


So, that's what I've done so far:
 - I added the Markdown-flavored renderer (`MarkdownTextRenderer`) and the Noop renderer (`NoopRenderer`) as well as abstract interfaces and factories for these two.
 - I implemented these into the list.
 - I wrote a lot of unit tests so nothing will (hopefully) break.

What needs to be done:
 - Fixing the regex (I don't know how to do that correctly. I'd appreciate someone's help.)
 - ~~Adding a ASCII escape code renderer.~~ (Done in a3f1384).
 - Tests under real work scenarios.
 - ~~Writing docs on how to implement it~~ (See: https://github.com/Matt3o12/termui/wiki/Color-Renderers)
 - ~~Refactoring the code (I need some advice from a native speaker here)[2] :)~~ (if you have any feedback, please give it now. Nothing is in the master branch, yet).
 - ~~Adding the renderers to other `Bufferer`-components?~~ Added in e9e3e40 (`Par` only).

--- 

**How do I use/test the new text renderer?** 
Just do: `myList.RendererFactory = termui.MarkdownTextRendererFactory{}` :)

**Are these changes backwards compatible?**
Short: yes, at least when you used `NewList()` instead of `List{}` as you're supposed to.

These changes add a new field to the List: RendererFactory, which generates the `TextRenderer`s. `List` uses this field for generating the appropriate (colorful) output. If you didn't use the list constructor (`NewList()`), the compiler will yield an error because of a missing field.  
If you use the constructor, the constructor will set the field to `NoopRendererFactory` which will not generate any formatted output (as the name states: `No Operation`).

--

If you have any question or feedback, just leave comment. 

---

[1]: The [current regex](https://github.com/Matt3o12/termui/blob/colored-list/textRender.go#L22) implementation for finding text sequences that need formatting, doesn't work properly. It will not format: `[[ERROR]](red,bold) foo` correctly, which should normalize to: `[ERROR] foo`. Here is the current regex in action with example it should find (https://regex101.com/r/fC0gD1/1). Currently only one expression fails. 
If you know a regex that will work, please let me know!

[2]: Everything that formats a text is called `TextRenderer`. In my dictionary I couldn't find Renderer and I wonder if that is a word, which makes sense. If you find that this word fits properly, let me know. If it doesn't, could you suggest a better one?

gizak commented:
Neat! There are a lot of changes I need to look into. Some suggestions:

1. For the ASCII escape code, you have to escape the escape code (e.g. `"\\e[33;My Text here"`). I am afraid that people won't like this limitation, so let us just focus on what we've done right now.

2. We just exposed `RendererFactory` to uses. This is part of implementations that users don't need to know. Suggest that it may be better if we add a field like `ColorSyntax="markdown"` instead.

3. To be honest when I saw `Noop`, the first thing got into my mind was: is that a noodle thing or something?... (Umm... IDK, `PlainRenderer` sounds better?)

Matt3o12 commented:
> For the ASCII escape code, you have to escape the escape code (e.g. "\\e[33;My Text here"). I am afraid that people won't like this limitation, so let us just focus on what we've done right now.

People can always choose their favorite (e.g. escape codes, the markdown-like syntax, or their own implementation). And if they really don't want to escape every character, they can just use raw strings (e.g. `` entry := `\e[33;My Text here` ``), which will work just fine. Of course, raw strings are not preferable for human text and have limitations but many people are used to escaping characters like `\` anyways when writing regular expressions in other languages.

Also, currently the markdown renderer can't render `[[ERROR]](red,bold) message` because the regex won't work correctly. This might be a problem for some projects and I don't know how to fix it (I'm not good at regexs). 

> We just exposed RendererFactory to uses. This is part of implementations that users don't need to know. Suggest that it may be better if we add a field like ColorSyntax="markdown" instead.

`user != developer`. The code aims at other developer who should be able to read docs and understand a few basic things about programming, like the [factory pattern](http://en.wikipedia.org/wiki/Factory_method_pattern). We could also add a method method like `UseColorSyntax(name string)` and `SetColorSyntax(factory TextRendererFactory)` just like I did in [themes.go](https://github.com/gizak/termui/blob/master/theme.go#L65).

> To be honest when I saw Noop, the first thing got into my mind was: is that a noodle thing or something?... (Umm... IDK, PlainRenderer sounds better?)

It was late and I couldn't come up with a better name. I'll rename it along with a  few other things :)

Matt3o12 commented:
2 questions still remain:

 - The regex doesn't quite work (do you know anyone who could fix that)?
 - Is TextRenderer a good name or do you know a better one.

Matt3o12 commented:
So, I think I'm done.  
I added 3 markdown renderes. The plain one (previously NoopRenderer), the MarkdownRenderer and the EscapeCodeRenderer (you don't even need to escape the `\033` because go recognizes it as a letter).   
I implemented the renderers in `Par` and `List`. The other components didn't seem to need it.  
I also wrote a [Wiki page](https://github.com/Matt3o12/termui/wiki/Color-Renderers). I'd be nice if you could merge it as well. The wiki is a separate git repo, if you didn't know, so this pull request does not include the page.

If you believe I've forgotten something, please let me know :)

TalkTakesTime commented:
@Matt3o12 
I'm not an expert by any means, but I think to have a robust and correct regex for what you want it to do you can't actually use Go's native regexp library. This is because Go uses the RE2 engine, which, while offering O(n) worse-case time, doesn't support quite a few features of PCRE such as recursion -- which would make what you want to do a lot simpler.

If you were to use a Go PCRE binding, the regex would probably look quite similar to
```
(\[((?:(?>[^[\]]+)|(?-2))*)\])\(([a-z\s,]+)\)
```
which gives results like [this](https://regex101.com/r/aN1kX1/4). This seems to be approximately how markdown works, from my experiments. If that's not what you're looking for, apologies, and let me know if I can improve it.

Also, for reference, [github.com/martinolsen/go-pcre/regexp](https://godoc.org/github.com/martinolsen/go-pcre/regexp) seems to be pretty much just a drop-in replacement for the standard `regexp` package, so it would be pretty easy to change over to using that.

Matt3o12 commented:
Hey!

Thank you for the suggestion! The regex seems to work (and I wrote a similar with Lookbehind which didn’t work either due to Go’s old regex implementation) but yours seem to be a better suited.
I’d be very happy to use that regex but I don’t know if adding another dependency is such a good idea.
Furthermore, `go get  github.com/martinolsen/go-pcre/regexp` fails:

	# github.com/martinolsen/go-pcre
	/Volumes/Shared HD/Developing/go/src/github.com/martinolsen/go-pcre/pcre.go:4:11: fatal error: 'pcre.h' file not found
	 #include <pcre.h>
	          ^
	1 error generated.

I don’t think we should add non-go dependencies. It makes cross-compiling harder and speed doesn’t really matter much when parsing short texts (and I think the time to [call a C function](http://lemire.me/blog/archives/2014/02/14/getting-good-performance-in-go-by-rewriting-parts-in-c/) is higher then just doing it in go).  
However, it is not up to me whether or not we should use that. 

BTW. If don’t really care whether or not it’s just a drop-in replacement as long as it returns where the regex starts and where it ends.

> On Apr 18, 2015, at 3:08 PM, TalkTakesTime <notifications@github.com> wrote:
> 
> @Matt3o12 <https://github.com/Matt3o12> 
> I'm not an expert by any means, but I think to have a robust and correct regex for what you want it to do you can't actually use Go's native regexp library. This is because Go uses the RE2 engine, which, while offering O(n) worse-case time, doesn't support quite a few features of PCRE such as recursion -- which would make what you want to do a lot simpler.
> 
> If you were to use a Go PCRE binding, the regex would probably look quite similar to
> 
> (\[((?:(?>[^[\]]+)|(?-2))*)\])\(([a-z\s,]+)\)
> which gives results like this <https://regex101.com/r/aN1kX1/4>. This seems to be approximately how markdown works, from my experiments. If that's not what you're looking for, apologies, and let me know if I can improve it.
> 
> Also, for reference, github.com/martinolsen/go-pcre/regexp <https://godoc.org/github.com/martinolsen/go-pcre/regexp> seems to be pretty much just a drop-in replacement for the standard regexp package, so it would be pretty easy to change over to using that.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/gizak/termui/pull/22#issuecomment-94164105>.
>

TalkTakesTime commented:
Hm, you're probably right. For the record, you can probably solve your `go get` issue by doing `brew install pcre` (I assume you're on a Mac due to `/Volumes`). Admittedly even having to do this is less than ideal.

Unfortunately, I believe it's impossible to correctly match an arbitrarily deep recursive expression using a single true regular expression (the language matched by PCRE is a superset of the language matched by "true" regular expressions, but I don't know its exact definition). If using an external dependency isn't an option, I guess the next best method for getting accurate results would be write a slightly more sophisticated parser than just a regex.

I'll have a look for alternatives and a play around myself and see if I come up with anything.

gizak commented:
Hi @Matt3o12, this is merged onto another branch `refactoring`, and I will use your code as a basic underlying text processing.

There is a lot of work need to be done for refactoring, but I think (hopefully) the branch will merge onto `master` soon.
